### PR TITLE
Create qchem refactoring feature branch

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
This PR creates a feature branch for the first phase of refactoring the quantum chemistry functionality in PennyLane. The details are discussed [here](https://docs.google.com/document/d/19B0M0KSoLppzjVopHJJeoez6uDCA4Qzfqrch7vcXf0Q/edit). In this PR, a differentiable equivalent of `qml.qchem` is created and can be accessed with `qml.qchem.hf`.

**Description of the Change:**
- The new `qchem.hf` subpackage is created.
- Functions for building a differentiable dipole moment observable are added to `qchem.hf.dipole`.
- Functions for building spin and particle number observables, with no external dependency, are added to `qchem.hf`.
- Modified versions of general-purpose functions needed for building observables are collected in `qchem.hf.observable`. 
- Modified version of functions for converting openfermion observables to pennylane ones are collected in `qchem.hf.convert`.
- The computational efficiency of the whole workflow is improved.

**Benefits:**
- The new functionality covers all features available in `qml.qchem`, is fully differentiable and is independent of external libraries. Adding this functions to `qml.qchem.hf` makes the future stepts of the quantum chemistry refactoring easier.

**Possible Drawbacks:**
Some codes will be duplicated for 1-2 release cycles. 

**Related GitHub Issues:**
NA

### CODE REVIEW

File: qchem/pennylane_qchem/__init__.py [hunk 1]
```diff
@@ -14,5 +14,5 @@
 """The PennyLane quantum chemistry package. Supports OpenFermion, PySCF,
 and Psi4 for quantum chemistry calculations using PennyLane.
 """
-from . import qchem
+from .qchem import hf
```
Comment (josh146) [inline qchem/pennylane_qchem/__init__.py:17]: are there any consequences from removing the `from . import qchem` line?

File: qchem/pennylane_qchem/qchem/hf/__init__.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This subpackage provides the functionality to perform differentiable
+Hartree-Fock calculations.
+"""
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/hf/__init__.py:17]: @soranjh I think you need to modify the `qml_qchem.rst` file to ensure these show up in the docs: https://github.com/PennyLaneAI/pennylane/blob/09e0992ee9bf5d885e601b043cd4b81c978b3c10/doc/code/qml_qchem.rst

File: qchem/pennylane_qchem/qchem/hf/convert.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    >>> ops = [
+    ...     qml.operation.Tensor(qml.PauliX(wires=['w0'])),
+    ...     qml.operation.Tensor(qml.PauliY(wires=['w0']), qml.PauliZ(wires=['w2']))
+    ... ]
+    >>> _pennylane_to_openfermion(coeffs, ops, wires=Wires(['w0', 'w1', 'w2']))
+    0.1 [X0] +
+    0.2 [Y0 Z2]
+    """
+    try:
+        import openfermion
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/hf/convert.py:207]: Don't forget to fix these issues before merging!

File: qchem/pennylane_qchem/qchem/hf/hamiltonian.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for computing the molecular Hamiltonian.
+"""
+# pylint: disable= too-many-branches, too-many-arguments, too-many-locals, too-many-nested-blocks
+import autograd.numpy as anp
+import pennylane as qml
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/hf/hamiltonian.py:19]: more pylint issues here (you can run pylint locally on this directory to show them all at once)

File: qchem/pennylane_qchem/qchem/hf/__init__.py (hunk 2 above)
Comment (josh146) [reply qchem/pennylane_qchem/qchem/hf/__init__.py:17]: Or do you not intend for this to show up in the docs?

**Review by josh146 — COMMENTED** [review]
@soranjh I did a quick skim through 🙂 No need to re-tag me for review.

Just wanted to double check: this is the first stage in moving HF to qchem and renaming some functions, but subsequent steps (moving functionality to new locations) is still to be done?

File: qchem/pennylane_qchem/__init__.py (hunk 1 above)
Comment (soranjh) [reply qchem/pennylane_qchem/__init__.py:17]: It is already changed to `from .qchem import hf`, not sure why it showed the older version.

File: qchem/pennylane_qchem/qchem/hf/__init__.py (hunk 2 above)
Comment (soranjh) [reply qchem/pennylane_qchem/qchem/hf/__init__.py:17]: I was thinking on of merging this feature branch now, before it gets even heavier, and then make further changes as direct PRs to master.

**Comment by soranjh** [comment]
> @soranjh I did a quick skim through slightly_smiling_face No need to re-tag me for review.
> 
> Just wanted to double check: this is the first stage in moving HF to qchem and renaming some functions, but subsequent steps (moving functionality to new locations) is still to be done?

Thanks @josh146. Yes, exactly, there will be few more PRs to fully polish qchem to its final format and then bring it inside pennylane. It is a bit easier for me to make those changes with separate PRs to master.

File: qchem/pennylane_qchem/qchem/hf/convert.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    except ImportError as Error:
+        raise ImportError("The OpenFermion package is required.") from Error
```
Comment (obliviateandsurrender) [inline qchem/pennylane_qchem/qchem/hf/convert.py:210]: Why do we need `from Error` here?

File: qchem/pennylane_qchem/qchem/hf/dipole.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+
+    where :math:`N_\mathrm{core}` is the number of core orbitals.
+
+    Args:
+        mol (Molecule): the molecule object
+        core (list[int]): indices of the core orbitals
+        active (list[int]): indices of the active orbitals
+
+    Returns:
+        function: function that computes the dipole moment integrals in the molecular orbital basis
```
Comment (obliviateandsurrender) [inline qchem/pennylane_qchem/qchem/hf/dipole.py:76]: Just a thought: 
Should we mention type for function as well? I think it is [Callable](https://www.python.org/dev/peps/pep-0483/#fundamental-building-blocks).

File: qchem/tests/hf/test_dipole.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Unit tests for functions needed for computing the dipole.
+"""
+import autograd
+import pennylane as qml
+import pytest
```
Comment (obliviateandsurrender) [inline qchem/tests/hf/test_dipole.py:19]: Do we worry about pylint warnings related to `third party import X should be placed before Y`?

**Review by obliviateandsurrender — APPROVED** [review]
Overall looks good to me @soranjh! I just left a few minor comments.

File: qchem/pennylane_qchem/qchem/hf/basis_data.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    "Be": 4,
+    "B": 5,
+    "C": 6,
+    "N": 7,
+    "O": 8,
+    "F": 9,
+    "Ne": 10,
+}
+
+STO3G = {
```
[the remark this answers is not in the document]
Comment (soranjh) [reply qchem/pennylane_qchem/qchem/hf/basis_data.py:33]: Yes, this is the plan.

File: qchem/pennylane_qchem/qchem/hf/convert.py (hunk 5 above)
Comment (soranjh) [reply qchem/pennylane_qchem/qchem/hf/convert.py:210]: To fix a pylint issue.

File: qchem/pennylane_qchem/qchem/hf/dipole.py (hunk 6 above)
Comment (soranjh) [reply qchem/pennylane_qchem/qchem/hf/dipole.py:76]: Maybe we should. I will consider it for a future PR.

File: qchem/tests/hf/test_dipole.py (hunk 7 above)
Comment (soranjh) [reply qchem/tests/hf/test_dipole.py:19]: I do not see such a warning, where is it exactly?

Comment (obliviateandsurrender) [reply qchem/tests/hf/test_dipole.py:19]: While running pylint locally, on lines with  `import pytest` I see the following: 
`C0411: third party import "import pytest" should be placed before "import pennylane as qml" (wrong-import-order)`

Comment (soranjh) [reply qchem/tests/hf/test_dipole.py:18]: Fixed, thanks.

**Review by ixfoduap — APPROVED** [review]

**Review by ixfoduap — APPROVED** [review]

**Pull request merged**